### ISSUE TITLE
DSpark speculative decoding: train Qwen3.6-27B-DSpark draft model

### ISSUE THREAD

**Issue by dbirks** [dbirks/home-k8s#50]
## What is DSpark?

DSpark (**D**eltaNet **Spark**) is a speculative decoding draft model framework from the `vllm-project/speculators` library. It extends the [DFlash](https://github.com/vllm-project/speculators) architecture with two additional heads that improve acceptance rates by ~34% over the baseline.

At inference, DSpark runs alongside the full verifier (Qwen3.6-27B) and proposes `N` draft tokens per step. The verifier confirms or rejects them in a single forward pass. When acceptance rates are high enough, this yields 4–7× throughput improvement with zero output quality degradation.

**Why train one for Qwen3.6-27B?**
- `z-lab/Qwen3.6-27B-DFlash` exists and works (already running in our vllm.yaml)
- No DSpark speculator for this model has been published yet (as of 2026-07-07)
- DSpark's Markov + confidence heads give an estimated additional ~1.5–2 tokens/step over DFlash
- The `speculators` v0.6.0 library fully supports the training pipeline
- We have the GPU (RTX 5090, SM120) and the model already cached on disk

---

## Architecture: Three-component stack

The draft model sits on top of a **frozen** verifier:

```
Qwen3.6-27B (frozen verifier)
      │
      │  hidden states at layers [1, 16, 31, 46, 61]
      ▼
┌─────────────────────────────────────────────────────┐
│  DFlash backbone (3 Qwen3 transformer layers)        │
│  FC projection: 5 × 5120 = 25600 → 5120             │
│  Anchored-block FlexAttention mask                  │
│  → draft logits                                     │
├─────────────────────────────────────────────────────┤
│  Markov head (rank=256 low-rank bigram)              │
│  W1[prev_token] @ W2^T → vocabulary bias            │
│  Conditions draft logits on last drafted token      │
├─────────────────────────────────────────────────────┤
│  Confidence head                                    │
│  Linear(hidden + Markov prev-emb → 1 scalar)        │
│  Predicts per-position acceptance probability       │
│  Gates dynamic draft length at inference            │
└─────────────────────────────────────────────────────┘
```

**Block-based drafting:** With `block_size=16`, each forward step produces up to 15 speculative tokens. A "block" has one anchor position (the real input token) plus 15 mask slots DSpark fills in.

**Knowledge distillation, not next-token prediction:** The training target is the verifier's full soft logit distribution at each position — not one-hot labels. This makes the draft learn to *mimic the verifier's distribution*, which directly minimizes rejection rate.

---

## How hidden state extraction works

`launch_vllm.py` starts the verifier with a special config that hooks into vLLM's forward pass:

```json
{
  "kv_connector": "ExampleHiddenStatesConnector",
  "kv_role": "kv_producer",
  "kv_connector_extra_config": {
    "shared_storage_path": "/output/data/hidden-states"
  }
}
```

When the trainer hits a training sample with no cached hidden states, it POSTs the conversation to the vLLM API (running on port 8001 alongside the trainer). vLLM runs the full 27B forward pass and saves the hidden states for layers 1, 16, 31, 46, 61 to a `.safetensors` file. The trainer then loads it and trains. With `--on-generate delete`, files are deleted after use (zero disk accumulation).

This "online" approach lets vLLM and the trainer share the GPU concurrently via CUDA time-slicing. The alternative ("offline") would require 500GB–3TB of disk for pre-cached hidden states — we don't have that.

⚠️ **Known conflict:** `ExampleHiddenStatesConnector` rejects `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` at startup. The training job script handles this: unset before `launch_vllm.py`, restore before `train.py`.

---

## Loss functions

```
L_total = compound_loss + confidence_head_alpha × conf_loss
```

| Loss | What it optimizes | Weight |
|------|-------------------|--------|
| TV (total variation) | `1 - acceptance_rate` directly | 0.9 |
| CE (cross-entropy) | Broad distribution alignment | 0.1 |
| BCE (confidence head) | Calibrated acceptance predictor | auto |

**Positional decay:** Earlier positions in each block get higher weight (γ=4.0). Position 1 gets full weight; position 7 gets ~0.22. This reflects that early speculative tokens have higher acceptance probability and contribute more to throughput.

---

### Prerequisites (already done ✅)
- [x] `token_freq.pt` computed and on `dspark-output` PVC
- [x] Arrow files for ShareGPT 5K samples on `dspark-output` PVC
- [x] vLLM image has `speculators` installed + scripts from git clone
- [x] `git` installed in image via `apt-get`

### Steps to kick off a training run

1. **Scale vLLM down** — training job needs the full GPU
   ```bash
   # In apps/vllm.yaml, set replicas: 0 then commit+push
   ```

2. **Activate the training job**
   ```bash
   # Rename .hold → .yaml
   mv apps/dspark-train-20260706.yaml.hold apps/dspark-train-20260706.yaml
   git add apps/dspark-train-20260706.yaml.hold apps/dspark-train-20260706.yaml
   git commit -m "feat(dspark): start training run"
   git push
   ```
   Flux picks it up and creates the K8s Job. The job script:
   - Installs `speculators` from git clone (gets `scripts/`)
   - Starts the verifier as a background vLLM API server
   - Runs `prepare_data.py` (skips if Arrow files exist)
   - Launches `train.py` with 5 epochs

3. **Monitor** — checkpoints every 0.25 epochs (≈ 4 per epoch, 20 total):
   ```bash
   kubectl logs -f job/dspark-train -n default
   ```
   Key metrics to watch: `accept_rate`, `train/loss`, `train/accept_len`

4. **Resume** — job is restartable. If it crashes, `train.py` detects the latest checkpoint in `/output/checkpoint/` and resumes automatically.

---

## Training configuration (apps/dspark-train-20260706.yaml.hold)

```
Model:          unsloth/Qwen3.6-27B-NVFP4
Dataset:        ShareGPT (5K samples, seq_len=8192)
Epochs:         5
LR:             3e-4
Draft layers:   3  (~1.4B params)
Block size:     16 (= 15 speculative tokens/step)
Markov rank:    256
Loss fn:        {"ce": 0.1, "tv": 0.9}
Target layers:  1 16 31 46 61
GPU memory:     vLLM 35% | trainer remainder
Checkpoint:     every 0.25 epoch → /output/checkpoint/
```

---

## Expected performance

**Acceptance rate at training milestones (based on RedHatAI GLM-5.2 DSpark numbers):**

| Epoch | Expected `accept_len` |
|-------|----------------------|
| 1 | ~2.5–3.5 |
| 3 | ~4.0+ ✅ (usable) |
| 5 | ~4.5–5.5 (target) |

**Throughput comparison (once trained), based on z-lab DFlash baseline for Qwen3.5-27B-DFlash at block_size=16:**

| Speculator | HumanEval | MBPP | GSM8K | Speedup |
|-----------|-----------|------|-------|---------|
| None (baseline) | 92 tok/s | 92 tok/s | 92 tok/s | 1× |
| z-lab DFlash | 572 tok/s | 495 tok/s | 461 tok/s | ~5× |
| DSpark (estimated) | ~700 tok/s | ~610 tok/s | ~560 tok/s | ~6–7× |

DSpark adds ~34% over DFlash via the Markov + confidence heads (per RedHatAI benchmarks on GLM).

---

## Time estimate

| Phase | Duration | Notes |
|-------|---------|-------|
| Data prep (`prepare_data.py`) | ~5 min | Already done ✅ — Arrow files persist on PVC |
| Hidden state gen for 5K samples (epoch 1) | ~2–3 hr | Online mode: vLLM forward pass per sample |
| Subsequent epochs | ~1–2 hr each | Faster if cached with `--on-generate cache` |
| Full 5-epoch run | **~8–14 hours** | On RTX 5090, online mode, 5K samples |
| Benchmark evaluation | ~30 min | `scripts/evaluate/evaluate.py throughput` |

**Total: 1–1.5 days** (mostly unattended GPU time)

To cut this to ~4 hours: use `--on-generate cache` (keep hidden states after epoch 1, reuse for epochs 2–5). Requires ~10–20GB free on `dspark-output` PVC.

---

## Post-training steps

1. **Upload to HuggingFace** (makes it reusable and shareable):
   ```bash
   # From inside the completed job pod or a debug pod:
   huggingface-cli upload david-birks/Qwen3.6-27B-DSpark /output/checkpoint/
   ```

2. **Wire into vllm.yaml:**
   ```yaml
   - "--speculative-config"
   - '{"method":"dspark","model":"david-birks/Qwen3.6-27B-DSpark","num_speculative_tokens":15}'
   ```

3. **Benchmark against DFlash baseline:**
   ```bash
   python scripts/evaluate/evaluate.py throughput \
     --target http://vllm.hoam.lan \
     --dataset RedHatAI/speculator_benchmarks \
     --output-dir ./benchmark_results
   ```
   Compare `acceptance_length` and `output_tps_median` against the current DFlash numbers.

4. **Iterate** — if `accept_len < 4.0` after epoch 3:
   - Try `--num-layers 5` (more draft capacity)
   - Try `--loss-fn '{"tv": 1.0}'` (pure TV, more aggressive)
   - Try a coding-focused dataset (`Magpie-Align/Magpie-Qwen3-Pro-200K-Filtered`)

---

## Known risks

- **Online mode instability:** vLLM and trainer compete for GPU memory under CUDA time-slicing. If the trainer OOMs, the job will crash and need to resume from the last checkpoint.
- **ExampleHiddenStatesConnector:** Uses shared filesystem for hidden states — `launch_vllm.py` must be on a writable path (`/output/data/hidden-states`). If the path is on a ReadOnly PVC mount the job will silently fail.
- **Exposure bias:** Markov head is trained with teacher forcing (ground-truth prev tokens) but at inference uses sampled draft tokens. `--use-off-policy-tokens` exists to mitigate this but adds training complexity.
- **No published DSpark for Qwen3.6:** This is first-party training, not fine-tuning a published checkpoint. Epoch 1–2 metrics will be lower than published baselines on other models.

---

## Task checklist

- [x] Fix issue #1: memory request (32Gi → 8Gi)
- [x] Fix issue #2: missing `scripts/` in wheel (apt-get install git + git clone)
- [x] Fix issue #3: `PYTORCH_CUDA_ALLOC_CONF` conflict with `ExampleHiddenStatesConnector`
- [x] Create no-GPU preflight job — validates apt+uv+git+prepare_data independently
- [x] Preflight job passes: `token_freq.pt` and Arrow files on `dspark-output` PVC
- [ ] Run first full training job (5 epochs, ShareGPT 5K)
- [ ] Upload checkpoint to HuggingFace
- [ ] Benchmark vs z-lab DFlash baseline
- [ ] (Optional) Train on coding dataset for better HumanEval numbers
- [ ] (Optional) Contribute checkpoint to `speculators` model hub

---

## Related

- #49 — Envoy Gateway for semantic routing (future: DSpark checkpoint could feed a specialized routing pool)
- `apps/dspark-train-20260706.yaml.hold` — training job definition
- `apps/dspark-preflight-20260707.yaml` — preflight job (completed)
- `DSPARK-INTERNALS.md` — deep-dive on the architecture
- `DSPARK-RUN-20260706.md` — run history and crash postmortems
- `z-lab/Qwen3.6-27B-DFlash` — DFlash baseline (currently in use in vllm.yaml)

**Comment by dbirks** [comment]
Closing as duplicate of #51, which has more comprehensive research and benchmarking details.
